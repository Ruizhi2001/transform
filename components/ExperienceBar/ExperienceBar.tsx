import { View, Text, Animated, Keyboard } from "react-native";
import React, { useEffect, useRef, useState } from "react";
import {
  DocumentReference,
  arrayUnion,
  doc,
  getDoc,
  updateDoc,
} from "firebase/firestore";
import { auth, db } from "../../FirebaseConfig";

type props = {
  height: number;
};

export const ExperienceBar: React.FC<props> = ({ height }) => {
  const animatedValue = useRef(new Animated.Value(-1000)).current;
  const reactive = useRef(new Animated.Value(-1000)).current;
  const [width, setWidth] = useState(0);
  const [exp, setExp] = useState(0);

  let user = auth.currentUser;
  let userDocRef: DocumentReference;

  if (user) {
    // get tasklist for user from firestore
    userDocRef = doc(db, "users", user.uid);
    getDoc(userDocRef).then((snap) => {
      if (snap.exists()) {
        setExp(snap.data().exp);
      } else {
        // snap.data() will be undefined in this case
        console.log("No such document!");
      }
    });
  }

  useEffect(() => {
    Animated.timing(animatedValue, {
      toValue: reactive,
      duration: 300,
      useNativeDriver: true,
    }).start();
  }, []);

  useEffect(() => {
    reactive.setValue(-width + (width * (exp % 100)) / 100);
  }, [exp, width]);

  return (
    <>
      <Text>LEVEL: {Math.floor(exp / 100)}</Text>
      <Text>{exp} EXP</Text>
      <View
        onLayout={(e) => {
          const newWidth = e.nativeEvent.layout.width;

          setWidth(newWidth);
        }}
        style={{
          height,
          backgroundColor: "white",
          borderRadius: height,
          overflow: "hidden",
        }}
      >
        <Animated.View
          style={{
            height,
            width: "100%",
            borderRadius: height,
            backgroundColor: "red",
            position: "absolute",
            left: 0,
            top: 0,
            transform: [
              {
                translateX: animatedValue,
              },
            ],
          }}
        />
      </View>
      <Text>⚡{100 - (exp % 100)} MORE EXP TO THE NEXT LEVEL⚡</Text>
    </>
  );
};
