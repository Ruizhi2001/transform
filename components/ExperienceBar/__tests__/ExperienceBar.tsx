import React from "react";
import renderer from "react-test-renderer";
import { ExperienceBar } from "../ExperienceBar";

// Mock the useAppSelector hook from redux
jest.mock("../../../redux/hooks.ts", () => ({
  useAppSelector: jest.fn((selector) => selector({ exp: 70 })), // Replace the '50' with a value for testing
}));

describe("ExperienceBar", () => {
  it("should match the snapshot", () => {
    const tree = renderer.create(<ExperienceBar height={40} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
