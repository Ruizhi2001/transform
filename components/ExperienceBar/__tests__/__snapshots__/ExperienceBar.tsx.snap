// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperienceBar should match the snapshot 1`] = `
[
  <Text>
    LEVEL: 
    NaN
  </Text>,
  <Text>
     EXP
  </Text>,
  <View
    onLayout={[Function]}
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 40,
        "height": 40,
        "overflow": "hidden",
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "red",
          "borderRadius": 40,
          "height": 40,
          "left": 0,
          "position": "absolute",
          "top": 0,
          "transform": [
            {
              "translateX": -1000,
            },
          ],
          "width": "100%",
        }
      }
    />
  </View>,
  <Text>
    ⚡
    NaN
     MORE EXP TO THE NEXT LEVEL⚡
  </Text>,
]
`;
